src/doc/style-guide/src/items.md

```diff
@@ -367,26 +367,49 @@ where
 ## Type aliases
 
 Keep type aliases on one line when they fit. If necessary to break the line, do
-so after the `=`, and block-indent the right-hand side:
+so before the `=`, and block-indent the right-hand side:
 
 ```rust
 pub type Foo = Bar<T>;
 
 // If multi-line is required
-type VeryLongType<T, U: SomeBound> =
-    AnEvenLongerType<T, U, Foo<T>>;
+type VeryLongType<T, U: SomeBound>
+    = AnEvenLongerType<T, U, Foo<T>>;
 ```
 
 Where possible avoid `where` clauses and keep type constraints inline. Where
-that is not possible split the line before and after the `where` clause (and
-split the `where` clause as normal), e.g.,
+that is not possible, prefer a trailing `where` clause over one that precedes
```

Comment:
Part of what was also done in #113380 was removing some (but not all) usages of "prefer".

My interpretation of this is that is indeed intentional, targeted more at the human developer, aware that the clause can go in either/both position, and trying to encourage the latter because we feel like the style guide has to provide rules for both positions?

Reply:
I don't think we should use "prefer" in any case where we expect a tool to do any kind of enforcement. But in this case, I don't think we expect a tool to convert between the two `where` positions?

Reply:
Tools should not convert. Prefix WC is an error in the positions where it matters, so maybe we don't even need to tell them about preference at all. Happy to reword or remove.

Reply:
Tools, including rustfmt, certainly _could_ and it's probably worth noting that during some of the earlier rfc and implementation discussions that was explicitly suggested/requested.

I know there's differing philosophies around the _should_ part of the question though

Reply:
Well given that I'm probably in the position of having the least context and caring the least about this specific wording, one of y'all are free to suggest  a rewording here.

Reply:
I personally feel both the text and the resultant behavior we're targeting are quite important. I think it's fully within t-style's domain to determine whether we want the default style to be free of clauses in the old position as part of the style guide prescriptions (there are other cases where prescriptions result in the formatter adding/removing tokens and/or moving things around)

And considering that we're pushing this independent of a style edition boundary I'd like to know what we're targeting for 2021 (even if e.g. we decide to not move/adjust for 2021, but defer deciding whether to prescribe that for 2024 to a later date)

Reply:
More concrete proposal of options I can think of:

* (A) we leave the text as-is, treat it as guidance to the developer, rustfmt has full rule/prescription coverage
* (B) only explicitly speak to the new position, don't mention the old position/don't frame the text in a "convert old to new" manner and instead just slightly adjust the text so that we on the rustfmt side can still point back to it as being implicitly applicable to the old position
* (C) explicitly prescribe the new position in a way that would require the tools to move a clause in the old position to the new.

My impression is that folks have reservations about option C so I'd be content to rule that out for now and/or punt to 2024 (or beyond). I also don't have any explicit objections to A (my original question was really just to verify my understanding of the intent), but I'd be willing to take a pass at suggesting some text for option B if anyone feels it'd be better to not explicitly mention/exemplify the clauses in the old position

Reply:
I think (A) is the best way to go here, or to just remove the wording altogether and leave it up to the language to enforce what to do with where clause placement. The guide could perhaps just explain what to do in the cases of prefix, suffix, and prefix+suffix where clauses and take no opinions.

Reply:
FWIW, I don't have objections to either (A) or (C) in isolation; however, I'd object to doing (C) *if we're not waiting for a new style edition*, because that would produce more churn.

As for (B), I think the style guide *should* be providing documentation for how to format any language construct that exists in the language. If something hasn't been removed from the language, we should document how to handle it (even if the handling is (C), "convert it to something else"). We shouldn't leave things undocumented/implied.

Reply:
OK, there's a bigger overriding consideration here. @compiler-errors just pointed out to me that the two locations have a *semantic difference*: `where` clauses before the `=` aren't enforced, while `where` clauses in the later position (which for item-level type aliases are currently only allowed in nightly) *are* enforced. So we should never move one to the other as a matter of mere *style*.

That means we should just describe both positions, and state that since there's a semantic difference between them, the Rust style does not express a preference or suggest moving constraints from one position to the other.

Reply:
Will update, which should resolve this convo. Hopefully after that we can kick off an FCP?